### DISCUSSION TITLE
Implementing stop loss and take profit orders

### DISCUSSION THREAD

**Discussion by sjhermanek** [blankly-finance/blankly#66] [Q&A] [answered]
Hi there - love the library!
I'm seeking your advise on implementing stop loss and take profit orders. What would be the most blankly-an way of doing so? Any advise would be greatly appreciated!
Thanks and keep up the amazing work!

**Comment by EmersonDove** [comment]
Hey I'm glad you like the package!

Currently we don't offer stop orders (besides a super secret stop limit in Coinbase Pro) although we really want to implement those! Binance gave us a lot of issues when we tried to implement it last time. I will add this to our backlog for our round 2 - we are just trying to develop features by demand right now. You can implement a very accurate stop limit engine if you're interested by using a websocket feed or and placing the limit at the activate price:

```python

import blankly


def new_price(price):
    print(price)
    # could check to see if any limits should be placed at this activate price


if __name__ == "__main__":
    manager = blankly.TickerManager('coinbase_pro', 'BTC-USD')
    manager.create_ticker(new_price)
```

You can also just rapidly query the price, or even use the `direct_calls` object to place stops on your exchange: 
`blankly.CoinbasePro().get_direct_calls().place_stop_order()`

**Comment by sjhermanek** [comment, marked as answer]
Sounds good. I'm excited for this feature to land in the future! In the meantime, I've decided to give this a bit of a go myself, aka a poor-man's version of it. 
The basic premise of my approach is to assemble a `book` of open positions from most recent `trade`'s, and pass those around in `variables`. Below, you can find some pseudo-code:

**Trades**
```
variables["trades"][state.time] = {DIRECTION_OF_TRADE, symbol, order_size, interface.get_price(symbol))
## DIRECTION_OF_TRADE will be in `BUY`, `SELL`, or `HOLD` in my logic.
```

**Book**
```
variables["book"][state.time] = {symbol, qty_held, last_trade, qty_considered, avg_acquisition_price}
## In here, I process trades from most recent to least recent, incrementing qty_held and (and internal) total spend on those trades; once qty_considered >= qty_held, I stop processing trades.
```

**Stop losses**
_I only trigger this if trade is either SELL or HOLD, since otherwise this would essentially just be a wash-sale. Trust the model, I guess ;)_
```
def stop_losses(variables, trade, cutoff):
## If last_price * cutoff < avg_acquisition_price: 
##   risk_managed_trade = "SELL"; 
## else:
##   risk_managed_trade = trade
```
Hope this might help you and others!

**Reply by EmersonDove** [reply]
Haha this is a pretty good solution - sorry to have to make you implement it. I will point others who ask here!

**Reply by sjhermanek** [reply]
Thanks! Want me to share the complete code?

**Reply by EmersonDove** [reply]
Sure, if you want to create an example repo with a requirements.txt and everything that would be really cool

**Comment by EmersonDove** [comment]
This feature has been implemented:

https://docs.blankly.finance/orders/take_profit_order
https://docs.blankly.finance/orders/stop_loss_order
